### PR TITLE
Initial version of adb-client

### PR DESCRIPTION
**Pull request by cstyan**
Just to start off, at the moment things break a lot.  I'll outline the issues I'm having further below.
I'm going to try and outline everything I've worked on and the thoughts I have so that hopefully any potentials issues I haven't seen are made visible to everyone else and so I can get some input on the direction you guys want this to go in.  

@jlipps @filmaj 

This is going to be a bit of a wall of text, sorry.

So far I have the following:

# adb-device
This is meant to act as an interface between a device and the client portion of the code base.  I would like to make the class an actual Interface, but I'm not sure if it's necessary so for now I'll just document what methods a concrete device class will need to implement.  The concrete device is a property of this class, and `adb-device` calls some of it's methods.

The constructor takes a connection type, ADB supports USB or TCP, obviously there will be some changes when we want to implement a TCP class but for now I've been working solely on USB.

# usb-device
Implements the required methods for communication with a device over USB.  At the moment I'm not happy with how the sequence of most of the methods has been written.  Unfortunately the flow of ADB isn't very nice to work with.  Rather than getting a useful error message back from the device when something fails we just get the actual failure message back.  When we send a command to the device it creates another process to act as a stream to us for that command.

Here's an example:
When we want to send the command `shell:ls.`, which is the equivalent of running `adb shell ls`, there are three things that can happen.  First, the initial creation of the new process can fail, but that doesn't really matter since the device will try again anyways and I've never gotten a failure message more than once.  Second, the process might be created just fine, but we'll still get some informational message back from the device rather than actual data we want.  This situation can occur after the initial failure message or on it's own.  Third, the very first message we get back is the actual data we requested.  Unfortunately we don't always seem to get all the data, but this could be an issue with how I'm reading from the USB interface itself.  I'll have to look into that more.  
The main issue here is that in order to react to the messages we receive and read again we'll have to do string comparisons, as well as determine what the potential messages could be for each command type.  There seems to be a pattern of the messages we don't actually want always starting with `adb`, but I still don't like having to do a string comparison on every incoming packet.  
The `_shell` method is a good example of the point at which one of those 3 scenarios from above could occur.  In `initConnection` and `handleAuth` you can see how each function is essentially hardcoded to fit the flow of the protocol.  I'm also not a fan of the `do { ... } while()` loop in `handleAuth`, is there a way to continuously do something until a condition is met with some of the new ES6 features?

# helpers & constants
Nothing special, just helper functions and some constants related to the ADB protocol that need to be used in multiple locations.

# OSX
I haven't been able to determine the cause, but there is a consistent issue with connecting to a device in OSX that doesn't show up in Linux.  If you run the current version of the `lib/examples/shell` file (the code to run the shell command is commented out), you can run it over and over on Linux and always get a pop up on the device asking you to authenticate.  In this case the entire handshake for ADB is working.  On OSX, more than half the time you're not able to complete a handshake with the device.  Sometimes the handshake blocks at the point when we send AUTH and our public key back to the device.  Sometimes it works every other time, sometimes it doesn't work at all.  Usually I need to reboot the device and give it a minute or so after being plugged in before trying to initiate the ADB connection.  Again, I'm not sure what the cause of this is.

# TL; DR
Reactionary processing of messages from the device will require string comparisons.  I think the way to go is to try and identify any pattern in the responses and turn the device classes into state machines.  It might also be necessary to use the streaming functionality of the [node-usb](https://github.com/nonolith/node-usb) library.  As far as the OSX problems, no clue how to proceed.  Throughout this I've been trying to find some answers in the code for Google's ADB implementation but haven't had much luck.  The `node-usb` library is just a wrapper for libusb, and I'm not sending or receiving any differently in OSX than Linux.  Any input is welcome regarding the direction to go in or any of the code, I'm not attached to the way I've built any of this.

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import USBDevice from './lib/usb-device';
```
Comment (sebv) [inline test.js]: missing `// transpile:main` (that will enable source map).

File: lib/examples/shell.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { connectToDevice } from '../../index.js';
```
Comment (sebv) [inline lib/examples/shell.js]: also // transpile:main on top.

**Comment by sebv** [comment]
> is there a way to continuously do something until a condition is met with some of the new ES6 features?

You mean you want to run something in background? You don't have to wait for it. You can do something like `B.resolve(handleAuth()).done();` instead of `await handleAuth()`

File: lib/adb-device.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import { CONNECTION_TYPES } from './constants';
+import USBDevice from './usb-device';
+
+// create one of these to interface with a device
+class ADBDevice {
```
Comment (sebv) [inline lib/adb-device.js]: I guess this is temporary, class hierarchy still needs improvement.

File: lib/constants.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+//TODO: detect these or command line arguement?
+export const ADB_KEY = {
+  PUBLIC_KEY:  ".android/adbkey.pub"
+, PRIVATE_KEY: ".android/adbkey"
+};
+
+export const CONNECTION_TYPES = {
+  USB: 0
```
Comment (sebv) [inline lib/constants.js]: Can use `USB=Symbol('usb')`

File: lib/helpers.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  for (let i = 0; i < buf.length; i++) {
+    crcResult = (crcResult + buf[i]) & 0xFFFFFFFF;
+  }
+  return crcResult;
+}
+
+// generates a message according to the ADB packet specifications
+// node that we don't append the actual data payload to the message
+// unless we're on tcp
+var generateMessage = function (cmd, arg1, arg2, payload, connectionType) {
```
Comment (sebv) [inline lib/helpers.js]: `function generateMessage`

File: lib/helpers.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  let payloadBuffer = (!Buffer.isBuffer(payload)) ? new Buffer(payload)
```
Comment (sebv) [inline lib/helpers.js]: should not need extra brackets.

File: lib/helpers.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  if (connectionType === CONNECTION_TYPES.TCP) {
+    console.log("writing data payload to packet");
+    payloadBuffer.copy(message, 24);
+  }
+  return message;
+};
+
+// takes a buffer from an inputEndpoint read and
+// creates the packet structure from the data
+var packetFromBuffer = function (buf) {
```
Comment (sebv) [inline lib/helpers.js]: function ...

File: lib/usb-device.js [hunk 8]
```diff
@@ -0,0 +1,238 @@
+// import constants
+import { USB_VENDOR_IDS, ADB_VALUES, CONNECT_VALUES,
+          ADB_COMMANDS, ADB_KEY } from './constants';
+let MAXDATA = CONNECT_VALUES.CONNECT_MAXDATA;
+// import helpers
+import { generateMessage } from './helpers';
+import { packetFromBuffer } from './helpers';
```
Comment (sebv) [inline lib/usb-device.js]: group the imports.

File: lib/usb-device.js [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+import fs from 'fs';
+import Promise from 'bluebird';
+import * as usb from 'usb';
+Promise.promisifyAll(usb);
+
+// local constants
+const homedir = (process.platform === 'win32') ? process.env.HOMEPATH
+                  : process.env.HOME;
+const keyPath = path.join(homedir, ADB_KEY.PUBLIC_KEY);
+const publicKeyString = fs.readFileSync(keyPath);
```
Comment (sebv) [inline lib/usb-device.js]: blocking sync method.

**Comment by sebv** [comment]
Also, this kind of decoding library need to be protected by unit tests.

File: lib/adb-device.js (hunk 3 above)
Comment (cstyan) [inline lib/adb-device.js]: I agree, I'd like the ADBDevice class to be an actual Interface that USBDevice and TCPDevice implement.

File: lib/constants.js (hunk 4 above)
Comment (cstyan) [inline lib/constants.js]: Do you mean I can do `export USB: Symbol('usb')` instead of exporting it as a key:value pair?

File: lib/usb-device.js (hunk 9 above)
Comment (cstyan) [inline lib/usb-device.js]: I'll change this to the async version and put some error checking around it, the adb key's are required for any of this to work.

**Comment by cstyan** [comment]
I want to make any major changes to the structure of the classes first, but I agree tests are necessary.  I don't have much experience with unit tests, especially for JS, so hopefully you guys can point me in the right direction there.

File: lib/constants.js (hunk 4 above)
Comment (sebv) [inline lib/constants.js]: Yeah, no big deal anyway.

File: test.js (hunk 1 above)
Comment (cstyan) [inline test.js]: I don't get how this works, is it the same as `#!/bin/bash`? I haven't seen this being used anywhere else.

**Comment by sebv** [comment]
We use http://mochajs.org/ (It's already configured in the boilerplate), and http://sinonjs.org/docs/

File: lib/usb-device.js (hunk 9 above)
Comment (cstyan) [inline lib/usb-device.js]: @sebv I think it would be better to move lines 17-20 to `constants.js`? Will `process.platform` work as part of an `export`?

Comment (sebv) [inline lib/usb-device.js]: Yes, that would just export the value.

Comment (cstyan) [inline lib/usb-device.js]: Okay, what about the `readFile`? I'm not sure what the best way is to export that value if we want to use the async version of the function as well as make sure the `readFile` returned something.  We can't even complete the handshake with a device if we can't get the adb keys.  I feel like using export and await together is a bad thing to do?

File: test.js (hunk 1 above)
Comment (sebv) [inline test.js]: https://github.com/appium/appium-gulp-plugins/blob/master/lib/transpiler.js#L33

File: lib/usb-device.js (hunk 9 above)
Comment (sebv) [inline lib/usb-device.js]: That needs to be a promise returning method. If you need to cache the result you can use `_.memoize`.

**Comment by cstyan** [comment]
Did some clean up and little fixes based on @sebv comments.  Hopefully someone else can jump in regarding the structure of the classes.  The ES6 class features seem to let you define a class and have another class extend that class, but there doesn't seem to be any sort of Abstract or Interface type, and I'm still fuzzy on JavaScripts prototypal inheritance.

File: index.js [hunk 10]
```diff
@@ -0,0 +1,7 @@
+import ADBDevice from './lib/adb-device';
+
+function connectToDevice (type) {
```
Comment (jlipps) [inline index.js]: what is the utility of this function? why wouldn't someone just run `new ADBDevice` on their own?

File: lib/adb-device.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  constructor(connectionType) {
```
Comment (jlipps) [inline lib/adb-device.js]: style nitpick: we do `functionname (args) {` (i.e., with spaces between)

File: lib/adb-device.js [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+      // errors yo
+      throw new Error("Invalid connection type.");
+    }
+  }
+
+  // initiates connection to the device and then
+  // registers for any events needed to get data
+  async start () {
+    await this.device.initConnection();
+    return;
```
Comment (jlipps) [inline lib/adb-device.js]: `return;` is the same as `return undefined;` which is the same as no return at all, so i'd remove this.

File: lib/adb-device.js (hunk 11 above)
Comment (cstyan) [inline lib/adb-device.js]: fixed, somehow I missed that one

File: lib/examples/shell.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  // , string: "ls"
+  // };
+  // await adb.open(command);
+  // console.log("shell opened");
+  console.log("closing device!");
+  await adb.closeConnection();
+}
+
+console.log("starting");
+start();
```
Comment (jlipps) [inline lib/examples/shell.js]: you might want to wrap this so any errors are output instead of swallowed in the promise rejection. the way that i usually do this is:

```js
import { asyncify } from 'asyncbox';
```

then

```js
asyncify(start);
```

It just handles the boilerplate of logging errors to the console if something in the promise chain is rejected

File: lib/helpers.js [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  }
+  return crcResult;
+}
+
+// generates a message according to the ADB packet specifications
+// node that we don't append the actual data payload to the message
+// unless we're on tcp
+function generateMessage (cmd, arg1, arg2, payload, connectionType) {
+  let payloadBuffer = !Buffer.isBuffer(payload) ? new Buffer(payload)
+                      : payload;
```
Comment (jlipps) [inline lib/helpers.js]: I would line the colon up under the question mark

File: lib/helpers.js [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+  let msgLength = payload !== null ? ADB_HEADER_LENGTH + payloadBuffer.length
+                  : ADB_HEADER_LENGTH;
```
Comment (jlipps) [inline lib/helpers.js]: same here (just some style nitpicks)

File: index.js (hunk 10 above)
Comment (cstyan) [inline index.js]: There's no reason they can't.  The whole exporting modules into an `index.js` file and then creating  another file to run the modules is strange to me coming from C/C++, so if I'm doing it incorrectly then please let me know.

File: lib/usb-device.js [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  // handles authentication between our machine and
+  // the android device upon initial connection
+  // TODO: attempt to sign token sent to us
+  async handleAuth () {
+    let publicKeyBuf = new Buffer(publicKeyString.length + 1);
+    publicKeyString.copy(publicKeyBuf);
+    publicKeyBuf[-1] = 0;
+    let authMsg = generateMessage(ADB_COMMANDS.CMD_AUTH
```
Comment (jlipps) [inline lib/usb-device.js]: style note: we'd write this as:

```js
let authMsg = generateMessage(ADB_COMMANDS.CMD_AUTH,
                              3, 0, publicKeyBuf);
```

i.e., when you need to break a line, line up the next argument just after the paren on the first line.

File: lib/usb-device.js [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+                                  , 0
+                                  , publicKeyBuf);
+    let command = null;
+    do {
+      await this._sendMsg(authMsg);
+      await this._sendMsg(publicKeyBuf);
+      console.log("Sent auth message and public key.");
+      let packet = await this._recvMsg(MAXDATA);
+      //TODO, actually check the fields of the packet
+      if(packet.command === ADB_COMMANDS.CMD_CNXN) {
```
Comment (jlipps) [inline lib/usb-device.js]: `if (`

File: lib/usb-device.js [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+  async handleAuth () {
+    let publicKeyBuf = new Buffer(publicKeyString.length + 1);
+    publicKeyString.copy(publicKeyBuf);
+    publicKeyBuf[-1] = 0;
+    let authMsg = generateMessage(ADB_COMMANDS.CMD_AUTH
+                                  , 3
+                                  , 0
+                                  , publicKeyBuf);
+    let command = null;
+    do {
```
Comment (jlipps) [inline lib/usb-device.js]: i love how simple this loop is with `async/await`

File: lib/usb-device.js [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+let MAXDATA = CONNECT_VALUES.CONNECT_MAXDATA;
+// import helpers
+import { generateMessage, packetFromBuffer } from './helpers';
+
+// required libraries
+import path from 'path';
+import fs from 'fs';
+import Promise from 'bluebird';
+import * as usb from 'usb';
+Promise.promisifyAll(usb);
```
Comment (jlipps) [inline lib/usb-device.js]: it's unclear why you're promisifying `usb`. AFAICT you don't `await usb.anything`. I took a brief brief look at the usb docs and they seemed to be sync functions (i.e., no need to await or promisify) but I could be wrong.

**Comment by jlipps** [comment]
@cstyan I did a first code pass without really evaluating the logic or the protocol (haven't had time to dig into that). looks like you've got the hang of async/await, with a few question marks around which functions are truly async or not (e.g. `usb` functions).

also, in general there's no reason to have a bare return unless it's for control flow (breaking out of something early). bare return is equivalent to no return in js.

**Comment by sebv** [comment]
There is no interface in interface, but you can easily do duck typing. Whether you create a base class or not does not really matter as long as the methods are protected by tests.

There are different solutions but for instance this is what you could do:

- Extract the constructor into a factory method (In general when constructor are throwing exception, it is a bad sign.).
- Remove the device property from ADBDevice and make it a base class, with empty method. 
- Make USBDevice extend ADBDevice and rename the corresponding methods.

File: index.js (hunk 10 above)
Comment (sebv) [inline index.js]: You can configure the main file in package.json and get rid of index.js if you want.

File: lib/usb-device.js (hunk 19 above)
Comment (cstyan) [inline lib/usb-device.js]: Initially, from reading the documentation about promisify, I thought calling promisifyAll on `usb` would promisify functions on the objects within a usb object, such as the endpoints.  Obviously that's not the case, so I can remove the promisify on `usb`.  I can only see one callback function on the usb object I might need to call anyways.

**Comment by cstyan** [comment]
Either way I'm going to have to take a day or two to learn how to do either duck typing or method testing properly.  I would prefer the tests route, but as you guys and the mobile team are going to be direct consumers of this code, whatever you think is going to be easier to use and maintain is the way we should go.  
My next immediate step is to identify any pattern or common prefix to device response that are from ADB and aren't the actual data we want.

File: LICENSE [hunk 20]
```diff
@@ -0,0 +1,13 @@
+Copyright 2012-2013 Appium Committers
```
Comment (filmaj) [inline LICENSE]: Probably want to update this :)

Comment (cstyan) [inline LICENSE]: @jlipps ^ this is from the `appium-package-master` creation process

**Comment by cstyan** [comment]
So the default Google implementation of ADB prints out these messages I was previously worried about having to read but not print, so that's helpful.  The messages we don't need to print are prefixed by `adb: ...` in the data portion of the packet.  I still think we might need to implement the streaming option from the `node-usb` library.

screen shot 2015-07-06 at 2 25 07 pm

**Comment by cstyan** [comment]
Okay I'm stupid, adb just always prints out all the data it gets sent in the `adb shell ls` case so some of the issues I was having were not really issues.  In the shell case we just keep reading packets from the device until we get a `CLSE` or some error.

**Comment by cstyan** [comment]
@jlipps @sebv I've finally got things set up to test properly.  The existing tests are for functions that don't really rely on a connection to a device.  I'm going to set up a fake device class and use `proxyquire` to test the `usb-device` class.  I'm hoping we can merge this PR sometime soon and move into smaller more focused branches, this ones getting pretty long.

File: lib/examples/reboot.js [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+  console.log("device started");
+  let command = {
+    type: "reboot"
+  };
+  await adb.open(command);
+  console.log("device should reboot now");
+}
+
+console.log("starting");
+start();
```
Comment (jlipps) [inline lib/examples/reboot.js]: this will swallow any errors, not sure if you care about that. i wrote a little util called `asyncify` in asyncbox that prints out errors.

```js
import { asyncify } from 'asyncbox';
...
asyncify(start);
```

File: lib/examples/shell.js (hunk 13 above)
Comment (jlipps) [inline lib/examples/shell.js]: oh lol forgot i had written this here already

File: lib/helpers.js [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+function findAdbDevice () {
+  console.log("trying to find a usb device");
+  let usbDevices = usb.getDeviceList();
+  for (let device of usbDevices) {
+    let deviceInterface = getAdbInterface(device);
+    if (deviceInterface !== null) {
+      console.log("Found an ADB device.");
+      return { device: device, deviceInterface: deviceInterface };
```
Comment (jlipps) [inline lib/helpers.js]: for regular js objects, we don't put spaces around the brackets. also, you can use es6 here!

```js
return {device, deviceInterface};
```

File: lib/helpers.js [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+    }
+  }
+  return null;
+}
+
+function getAdbInterface (device) {
+  device.open();
+  if(device.interfaces === null) { return null; }
```
Comment (jlipps) [inline lib/helpers.js]: `if (`

File: lib/usb-device.js [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+import { generateMessage, packetFromBuffer } from './helpers';
+
+// required libraries
+import path from 'path';
+import fs from 'fs';
+import Promise from 'bluebird';
+
+// local constants
+const homedir = process.platform === 'win32' ? process.env.HOMEPATH
+                  : process.env.HOME;
```
Comment (jlipps) [inline lib/usb-device.js]: let's line up `:` underneath `?`

File: lib/usb-device.js [hunk 25]
```diff
@@ -0,0 +1,210 @@
...
+    }
+  }
+
+  // sends a message, checks that message is a Buffer
+  async _sendMsg (msg) {
+    // this error checking breaks the function :/
+    // if(typeof msg !== Buffer) {
+    //   throw new Error("Message was not of type Buffer.");
+    // }
+    return await this.outputEndpoint.transferAsync(msg);
```
Comment (jlipps) [inline lib/usb-device.js]: `return await ` is actually the same as just `return`

in the first, we get a value from a promise and wrap it with a promise. in the second, we return the promise directly. so we've been preferring the second method

File: lib/usb-device.js [hunk 26]
```diff
@@ -0,0 +1,210 @@
...
+        dataLen = dataLen - readData.length;
+        packet.data = Buffer.concat([packet.data, readData]);
+      }
+    }
+    return packet;
+  }
+
+  async _shell (commandString) {
+    console.log("inside shell function");
+    let shellString = "shell:" + commandString + ".";
```
Comment (jlipps) [inline lib/usb-device.js]: we usually use interpolated strings

File: lib/usb-device.js [hunk 27]
```diff
@@ -0,0 +1,210 @@
...
+    let clseMsg = generateMessage(ADB_COMMANDS.CMD_CLSE
+                              , localId
+                              , remoteId
+                              , "");
+    await this._sendMsg(clseMsg);
+    console.log("sent CLSE");
+  }
+
+  async _reboot() {
+    let rebootString = "reboot:" + ".";
```
Comment (jlipps) [inline lib/usb-device.js]: seems like you don't need the concatenation operator here :-)

**Comment by jlipps** [comment]
just did another pass; looking good!

File: lib/usb-device.js (hunk 25 above)
Comment (cstyan) [inline lib/usb-device.js]: without the await will it not just return immediately with `undefined`?

File: lib/examples/reboot.js (hunk 21 above)
Comment (cstyan) [inline lib/examples/reboot.js]: I'll add this in.  Going forward the code is going to need some clean up and more error handling so `asyncify` will be handy.

Comment (cstyan) [inline lib/examples/reboot.js]: just noticed that this comment was for one of the usage examples, I think adding better error handling is the way to go

**Comment by cstyan** [comment]
There's a couple issues currently around timeouts on the libusb transfers, different behaviour in Linux vs. OSX.  I have a small example I'll put up in a repo.

**Comment by cstyan** [comment]
@jlipps @sebv I'm curious what you guys think of the current class structure.  I'm starting to run into issues where certain functionality should be available to a USB or TCP device but instead I just end up writing it straight into the USB device class, and that bothers me.  

I think earlier in the process Sebv mentioned writing tests to ensure child classes implement required functions, is that a better option than the current structure?  Maybe this shouldn't even be an issue, since we don't really need adb over TCP?

**Comment by cstyan** [comment]
@jlipps if you don't have any objections, I'm going to merge this PR

As far as I can tell the only command you guys are using that I haven't implemented is the port forwarding

**Comment by jlipps** [comment]
@cstyan go for it; sorry I never got time to do a full review

**Comment by cstyan** [comment]
@jlipps no worries, can always go over that in the future at some point

merging
